cpp/comms/std/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,408 @@
+  // Not supported by NCCL
+  ASSERT(
+    false,
+    "ERROR: commSplit called but not supported in this comms implementation.");
+}
+
+void cumlStdCommunicator_impl::barrier() const {
+  CUDA_CHECK(cudaMemsetAsync(sendbuff, 0, sizeof(int), stream));
+  CUDA_CHECK(cudaMemsetAsync(recvbuff, 0, sizeof(int), stream));
+
+  allreduce(sendbuff, recvbuff, 1, MLCommon::cumlCommunicator::INT,
+            MLCommon::cumlCommunicator::SUM, stream);
+
+  cudaStreamSynchronize(stream);
+}
+
+void cumlStdCommunicator_impl::isend(const void *buf, int size, int dest,
+                                     int tag, request_t *request) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  request_t req_id;
+  if (_free_requests.empty())
+    req_id = _next_request_id++;
+  else {
+    auto it = _free_requests.begin();
+    req_id = *it;
+    _free_requests.erase(it);
+  }
+
+  ucp_tag_t ucp_tag = (ucp_tag_t)tag;
+
+  ucp_ep_h *ep_arr = *_ucp_eps;
+  ucp_ep_h ep_ptr = ep_arr[dest];
+
+  struct ucx_context *ucp_request = ucp_isend(ep_ptr, buf, size, tag);
+
+  _requests_in_flight.insert(std::make_pair(req_id, ucp_request));
+  *request = req_id;
+
+  ucs_status_t flush_status = flush_ep(_ucp_worker, ep_ptr);
+#endif
+
+  ASSERT(UCX_ENABLED, "cuML Comms not built with UCX support");
+}
+
+void cumlStdCommunicator_impl::irecv(void *buf, int size, int source, int tag,
+                                     request_t *request) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  request_t req_id;
+  if (_free_requests.empty())
+    req_id = _next_request_id++;
+  else {
+    auto it = _free_requests.begin();
+    req_id = *it;
+    _free_requests.erase(it);
+  }
+
+  ucp_ep_h *ep_arr = *_ucp_eps;
+  ucp_ep_h ep_ptr = ep_arr[source];
+
+  struct ucx_context *ucp_request =
+    ucp_irecv(_ucp_worker, ep_ptr, buf, size, tag);
+
+  _requests_in_flight.insert(std::make_pair(req_id, ucp_request));
+  *request = req_id;
+#endif
+
+  ASSERT(UCX_ENABLED, "cuML Comms not built with UCX support");
+}
+
+void cumlStdCommunicator_impl::waitall(int count,
+                                       request_t array_of_requests[]) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  std::vector<struct ucx_context *> requests;
+  requests.reserve(count);
+
+  for (int i = 0; i < count; ++i) {
+    auto req_it = _requests_in_flight.find(array_of_requests[i]);
+    ASSERT(_requests_in_flight.end() != req_it,
+           "ERROR: waitall on invalid request: %d", array_of_requests[i]);
+    requests.push_back(req_it->second);
+    _free_requests.insert(req_it->first);
+    _requests_in_flight.erase(req_it);
+  }
+
+  int done = 0;
+  for (struct ucx_context *req : requests) {
+
+    ASSERT(req != nullptr, "Request %d should not be null");
+
+    wait(_ucp_worker, req);
```

Comment:
Is it not better to have a non-blocking check? This way you could go through a list of requests, checking (as opposed to waiting for them to complete) until the list is empty (removing requests as they complete).

Reply:
That is true. My thinking was that since `waitall` is meant to block anyways, we could just force the blocking with `wait` but it does not need to be done this way.

Reply:
Done.